cmd/start.go

```diff
@@ -110,28 +113,70 @@ func runStartCmd(cmd *cobra.Command, args []string) {
 	if err != nil {
 		glog.Fatalf("error creating clients: %v", err)
 	}
-	stopCh := make(chan struct{})
-	run := func(stop <-chan struct{}) {
 
-		ctx := createControllerContext(cb, startOpts.name, stopCh)
-		if err := startControllers(ctx); err != nil {
-			glog.Fatalf("error starting controllers: %v", err)
+	// TODO: Kube 1.14 will contain a ReleaseOnCancel boolean on
+	//   LeaderElectionConfig that allows us to have the lock code
+	//   release the lease when this context is cancelled. At that
+	//   time we can remove our changes to OnStartedLeading.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		sig := <-ch
+		glog.Infof("Shutting down due to %s", sig)
+		cancel()
+
+		// exit after 2s no matter what
+		select {
+		case <-time.After(2 * time.Second):
+			glog.Fatalf("Exiting")
+		case <-ch:
+			glog.Fatalf("Received shutdown signal twice, exiting")
 		}
+	}()
 
-		ctx.CVInformerFactory.Start(ctx.Stop)
-		ctx.InformerFactory.Start(ctx.Stop)
-		close(ctx.InformersStarted)
-
-		select {}
-	}
-
+	// TODO: when we switch to graceful lock shutdown, this can be
+	// moved back inside RunOrDie
+	lock := createResourceLock(cb)
 	leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
-		Lock:          createResourceLock(cb),
+		Lock:          lock,
 		LeaseDuration: leaseDuration,
 		RenewDeadline: renewDeadline,
 		RetryPeriod:   retryPeriod,
 		Callbacks: leaderelection.LeaderCallbacks{
-			OnStartedLeading: run,
+			OnStartedLeading: func(stop <-chan struct{}) {
+				controllerCtx := createControllerContext(cb, startOpts.name, ctx.Done())
+				if err := startControllers(controllerCtx); err != nil {
+					glog.Fatalf("error starting controllers: %v", err)
+				}
+
+				controllerCtx.CVInformerFactory.Start(ctx.Done())
+				controllerCtx.InformerFactory.Start(ctx.Done())
+				close(controllerCtx.InformersStarted)
+
+				select {
+				case <-ctx.Done():
+					// WARNING: this is not completely safe until we have Kube 1.14 and ReleaseOnCancel
+					//   and client-go ContextCancelable, which allows us to block new API requests before
+					//   we step down. However, the CVO isn't that sensitive to races and can tolerate
+					//   brief overlap.
+					glog.Infof("Stepping down as leader")
+					// give the controllers some time to shut down
+					time.Sleep(100 * time.Millisecond)
+					// if we still hold the leader lease, clear the owner identity (other lease watchers
+					// still have to wait for expiration) like the new ReleaseOnCancel code will do.
+					if err := lock.Update(resourcelock.LeaderElectionRecord{}); err == nil {
+						// if we successfully clear the owner identity, we can safely delete the record
+						if err := lock.Client.ConfigMaps(lock.ConfigMapMeta.Namespace).Delete(lock.ConfigMapMeta.Name, nil); err != nil {
```

Comment:
Why delete the config map at all?

Reply:
Deleting the config map is the only way to trigger a new election today.

Reply:
Hmm, okay